### ISSUE TITLE
`--locked` does not work with `cargo publish` or `cargo package`

### ISSUE THREAD

**Issue by lolbinarycat** [rust-lang/cargo#13986]
### Problem

updating a package's version in Cargo.toml, running `cargo publish`, then creating a tagged commit will cause a desync between the commit and the version on crates.io in the lockfile.

the crates.io version will have the appropriate version in Cargo.lock, but the tagged commit will have the wrong version info in the lockfile, causing problems for anyone who tries to build that version with cargo build --locked

### Steps

1. make a package in a git repo
2. make an init commit with everything checked in
3. change the package.version field in Cargo.toml
4. run cargo publish
5. make a new tagged commit
6. run `cargo build --locked`

### Possible Solution(s)

make cargo publish refuse to publish if the lockfile is out of date

### Notes

this is particularly annoying when trying to package a rust program with nix, often the only solution is to vendor a fixed lockfile

### Version

```text
release: 1.80.0-nightly
commit-hash: a8d72c675ee52dd57f0d8f2bae6655913c15b2fb
commit-date: 2024-05-24
host: x86_64-unknown-linux-gnu
libgit2: 1.7.2 (sys:0.18.3 vendored)
libcurl: 8.6.0-DEV (sys:0.4.72+curl-8.6.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: NixOS 23.11.0 [64-bit]
```

**Comment by epage** [comment]
To double check my understand, `cargo publish` is updating the lockfile on disk as part of its operation.  Does `cargo publish` fail, saying that the filesystem is dirty and you need to pass `--allow-dirty`?

**Comment by lolbinarycat** [comment]
no, `cargo publish` succeeds, and publishes a different version of the lockfile than the one that exists on disk.

**Comment by epage** [comment]
Ah, I was looking at the wrong layer of abstraction.  `cargo package` calls into `package` which updates the lockfile but `cargo publish` calls into `package_one` which is after that lockfile update.

```console
$ cargo publish -n
   Compiling cargo v0.81.0 (/home/epage/src/personal/cargo)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 10.03s
     Running `/home/epage/src/personal/cargo/target/debug/cargo -Zscript publish -n`
    Updating crates.io index
warning: manifest has no description, license, license-file, documentation, homepage or repository.
See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for more info.
   Packaging cargo-13986 v0.1.1 (/home/epage/src/personal/dump/cargo-13986)
   Verifying cargo-13986 v0.1.1 (/home/epage/src/personal/dump/cargo-13986)
   Compiling cargo-13986 v0.1.1 (/home/epage/src/personal/dump/cargo-13986/target/package/cargo-13986-0.1.1)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.26s
    Packaged 6 files, 1.1KiB (857.0B compressed)
   Uploading cargo-13986 v0.1.1 (/home/epage/src/personal/dump/cargo-13986)
warning: aborting upload due to dry run

$ cargo package
warning: manifest has no description, license, license-file, documentation, homepage or repository.
See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for more info.
error: 1 files in the working directory contain changes that were not yet committed into git:

Cargo.lock

to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag
```

**Comment by epage** [comment]
I worry that there are shenanigans that people do during publish such that having `cargo publish` also generate the lockfile would be a breaking change.

I think changing the behavior in the presence of `--locked` would be reasonable.  If people are relying on that, then they are going counter to the documentation.  Still not a great answer because you have to know to use it.

For myself, after doing enough releases, I feel like `cargo publish` is best left to plumbing and people should be using a higher level release workflow as talked about at https://doc.rust-lang.org/cargo/reference/publishing.html#publishing-a-new-version-of-an-existing-crate

**Comment by ehuss** [comment]
The Cargo team briefly looked at this today, and noted that `cargo publish --locked` and `cargo package --locked` doesn't work today (`--locked` seems to be ignored). That would probably good to fix, though it will require specialized logic since the packaging operation rewrites the lock file. One idea is to just run resolution on the entire workspace, and ensure that it is locked. (An alternative would be to have some logic that verifies that the new lockfile is a true subset, but we probably don't want to or need to do that.)

The default behavior is unlikely something we could change (at least without an edition), since people may be publishing in repos that do not check in `Cargo.lock`.

**Comment by mattc1170** [comment]
Was there a fix for this issue committed to Cargo 1.83.0? The publish workflow for one of my team's workspaces is now broken when using the latest stable toolchain. Our CI workflow has been using `cargo publish --locked` and was working fine until 1.83.0, where we started seeing the following types of errors:

`error: the lock file /build/pixel-base/Cargo.lock needs to be updated but --locked was passed to prevent this`

I suppose we can work around this problem simply by removing `--locked` from our invocation of cargo publish since it wasn't doing anything before in the first place. :-)

**Comment by ibilalkayy** [comment]
I noticed this behavior by changing the version in Cargo.toml file and after executing the `cargo build --locked`, it gives me this error:

> terminal$: cargo build --locked
error: the lock file /home/ibilalkayy/Documents/rust/cargo_testings/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --locked flag and use --offline instead.

But when I only write the `cargo—- build` command without writing the—-locked flag, the version changes successfully, and it becomes easy to publish.

Now I want to know if the --locked flag should be removed or something needs to be changed?

Thank you!

**Comment by ibilalkayy** [comment]
Hey @lolbinarycat  here is the behavior I noticed but before this, let me tell you what I did run.

### First step

```
cargo init .
cargo build

git init
git add .
git commit -m "Initial commit for my_package"
git remote add origin <repo_url>

cargo login
cargo publish

git tag v1.0.0
git push origin v1.0.0

```
### Second step

#### Update the version in the Cargo.toml file

> [package]
> name = "my_package"
> version = "0.2.0"

```
cargo build

git add Cargo.toml
git commit -m "Update version to 0.2.0"
git tag v0.2.0
git push origin v0.2.0

cargo publish

```

### Check the tagged commit

```
git clone <your-repo-url>
cd my_package

git checkout v0.2.0

cargo build --locked
```

It will give the error message.

If I add the `Cargo.lock` in the git also while updating the version, this error won't occur

**Comment by epage** [comment]
The issue asks for `cargo package` / `cargo publish` to always be locked.  Looks like `--locked` isnt' working and we want that at least fixed.

The reproduction steps should then be

1. `cargo init`
2. `cargo check`
3. `git commit -a` (for easy reverting)
4. Change version in `Cargo.toml`
5. `cargo package --locked`

Expected: `cargo package` should fail

Actual: `cargo package` passes

**Comment by epage** [comment]
Trying out those steps, it looks like `publish --locked` and `package --locked` do fail

@lolbinarycat can you verify that passing `--locked` does what you are expecting?  Is there a reason this isn't sufficient for your use case?

**Comment by lolbinarycat** [comment]
it sounds like `--locked` does as close as possible to what i expected.

**Comment by cceckman-at-fastly** [comment]
👋 I've been struggling with something that rhymes with this in https://github.com/fastly/Viceroy/pull/506.

If we regenerate lockfiles, we'll pick up `spdx` version `0.10.9`, which requires Rust 1.85. If we upgrade to Rust 1.85, something in our test suite panics on Windows (I haven't dug further than that.) (`rust-toolchain` in this repository is 1.83.)

So: we want to publish the public crates (`viceroy-lib` under `/lib` and `viceroy` under `/cli`) using _exactly_ the checked-in lockfiles.

I'm not able to do this. If I run `cargo publish` _or_ `cargo publish --locked`, Cargo attempts to regenerate the lockfiles - picks up the upgraded `spdx`, and then fails.

To my understanding, Cargo shouldn't be changing the lockfiles if it is passed `--locked`, yes?

**Comment by epage** [comment]
If you have an existing, up-to-date lockfile, `cargo package` should reuse all of your registry dependencies. Something more is likely going on. Probably best for this to be a dedicated issue as this appears unrelated to this one.
